### ISSUE TITLE
Incorrect Congnitive Complexity value

### ISSUE THREAD

**Issue by Allemweich** [rust-lang/rust-clippy#11164]
### Summary

I tried the cognitive_complexity lint on a basic code example. The result does not match up.

Input:
```rust
fn main() -> Result<()> {
    for _i in 0..10 { // +1
        for _k in 0..10 { // +2
            for _l in 0..10 { // +3
                for _m in 0..10 { // +4
                    for _n in 0..10 { // +5
                        for _o in 0..10 { // +6
                            print!("")
                        }
                    }
                }
            }
        }
    }

    Ok(())
}
```

**Expected result**: 21
**Actual Result**: 7

**Additional Details**
To my understanding, the cognitive complexty of the main function is 21, as each for loop adds a value of 1, plus 1 for each level of nesting that it is already in.

However the warning I get says that it is 7:
```
warning: the function has a cognitive complexity of (7/1)
  --> src\main.rs:17:4
   |
17 | fn main() -> Result<()> {
   |    ^^^^
   |
note: the lint level is defined here
  --> src\main.rs:1:9
   |
1  | #![warn(clippy::cognitive_complexity)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: you could split it up into multiple smaller functions
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cognitive_complexity
```

### Reproducer

I tried this code:
```rust
fn main() -> Result<()> {
    for _i in 0..10 { // +1
        for _k in 0..10 { // +2
            for _l in 0..10 { // +3
                for _m in 0..10 { // +4
                    for _n in 0..10 { // +5
                        for _o in 0..10 { // +6
                            print!("")
                        }
                    }
                }
            }
        }
    }

    Ok(())
}
```

I expected to see this happen:
warning: the function has a cognitive complexity of (21/1)

Instead, this happened:
warning: the function has a cognitive complexity of (7/1)

### Version

```text
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-pc-windows-msvc
release: 1.65.0
LLVM version: 15.0.0
```

**Comment by Allemweich** [comment]
I found the following test (comments added by me).
```rust
fn kaboom() {
    let n = 0;
    'a: for i in 0..20 { // +1 (nesting level 0)
        'b: for j in i..20 { // +2 (nesting level 1)
            for k in j..20 { // +3 (nesting level 2)
                if k == 5 { // +4 (nesting level 3)
                    break 'b;
                }
                if j == 3 && k == 6 { // +5 (nesting level 3, +1 for introducting a boolean operation)
                    continue 'a;
                }
                if k == j { // +4 (nesting level 3)
                    continue;
                }
                println!("bake");
            }
        }
        println!("cake");
    }
    
    // Cognitive Complexity: 20
}
```
By counting I arrive at a value of 20, but the test suite asserts (and measures) a value of 7.
Clearly there is some mismatch of definitions here. I go by the original specification https://www.sonarsource.com/docs/CognitiveComplexity.pdf

**Comment by felix91gr** [comment]
@epic-64 yeah, sorry about that. I was the last to touch that lint.

In a nutshell: the lint is broken and the entire concept of linting "Cognitive Complexity" is nonsense. Don't believe what SonarSource says about its marvels; it's just not true, fam. That's how they got me.

It's a lint that cannot be accomplished with modern techniques. It's a good few years away, perhaps a decade or more.

https://github.com/rust-lang/rust-clippy/issues/3793 Here you can see the logs of my investigation into it.

---

@flip1995 is there a way to archive this lint further down than Nursery? Maybe even... deleting it? There's at least a handful of users who've tried it and I don't think it would be honest to implement SonarSource's marketing whitepaper and call it a day.
